libc/unistd_file.c

```diff
@@ -0,0 +1,774 @@
+		TEST_ASSERT_EMPTY(buf);
+	}
+	TEST_ASSERT_EQUAL_INT(0, close(fdr));
+	TEST_ASSERT_EQUAL_INT(0, close(fd2));
+}
+
+
+/* test simultaneous write() on duplicated file descriptor */
+TEST(unistd_file, file_write_dup)
+{
+	int sum;
+
+	/* duplicate file descriptor and open another for reading */
+	fd2 = dup(fd);
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(0, fd2);
+	int fdr = open(filename, O_RDONLY);
+	TEST_ASSERT_GREATER_OR_EQUAL(0, fdr);
+	{
+		/* write to the same file consecutively from descriptor and its copy */
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, write(fd, LINE1, sizeof(LINE1) - 1));
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE2) - 1, write(fd2, LINE2, sizeof(LINE2) - 1));
+
+		/* try to read too much for simple error detection */
+		sum = (sizeof(LINE1) - 1) + (sizeof(LINE2) - 1);
+		TEST_ASSERT_EQUAL_INT(sum, read(fdr, buf, sum));
+		TEST_ASSERT_EQUAL_STRING_LEN(LINE1, &buf[0], sizeof(LINE1) - 1);
+		TEST_ASSERT_EQUAL_STRING(LINE2, &buf[sizeof(LINE1) - 1]);
+	}
+}
+
+/* Test write() to closed descriptor */
+TEST(unistd_file, file_readwrite_badfd)
+{
+	TEST_ASSERT_EQUAL_INT(0, close(fd));
+
+	TEST_ASSERT_EQUAL_INT(-1, write(fd, LINE1, sizeof(LINE1) - 1));
+	TEST_ASSERT_EQUAL_INT(-1, read(fd, buf, sizeof(LINE1) - 1));
+
+	/* this test closes fd by itself, invalidate it */
+	fd = -1;
+}
+
+/* test write() outside of the file */
+TEST(unistd_file, file_write_efbig)
+{
+	/* set offset in file at some positive value */
+	TEST_ASSERT_GREATER_OR_EQUAL_INT(0, lseek(fd, 1, SEEK_SET));
+	TEST_ASSERT_EQUAL_INT(-1, write(fd, LINE2, sizeof(LINE2)));
+	TEST_ASSERT_EQUAL_INT(EFBIG, errno);
+}
+
+/* test write on readonly file descriptor */
+TEST(unistd_file, file_write_readonly)
+{
+	fd2 = open(filename, O_RDONLY);
+	TEST_ASSERT_GREATER_OR_EQUAL(0, fd2);
+	{
+		TEST_ASSERT_EQUAL_INT(-1, write(fd2, LINE1, sizeof(LINE1) - 1));
+		TEST_ASSERT_EQUAL_INT(EBADF, errno);
+	}
+	TEST_ASSERT_EQUAL_INT(0, close(fd2));
+}
+
+/* test simple pipe() usage */
+TEST(unistd_file, file_readwrite_pipe)
+{
+	int p[2];
+
+	if (pipe(p) != 0) {
+		/* ignore test if not pipe() not supported (errno == ENOSYS) */
+		TEST_ASSERT_EQUAL_INT(ENOSYS, errno);
+		TEST_IGNORE();
+	}
+	else {
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, write(p[1], LINE1, sizeof(LINE1) - 1));
+		TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, read(p[0], buf, sizeof(LINE1) - 1));
+
+		TEST_ASSERT_EQUAL_INT(0, close(p[0]));
+		TEST_ASSERT_EQUAL_INT(0, close(p[1]));
+	}
+}
+
+/* test simple use of lseek() */
+TEST(unistd_file, file_lseek)
+{
+	/* fill file with content */
+	TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, write(fd, LINE1, sizeof(LINE1) - 1));
+
+	/* fallback to absolute beginning */
+	TEST_ASSERT_EQUAL_INT(0, lseek(fd, (off_t)0, SEEK_SET));
+	/* read line */
+	TEST_ASSERT_EQUAL_INT(sizeof(LINE1) - 1, read(fd, buf, sizeof(LINE1) - 1));
+	TEST_ASSERT_EQUAL_STRING(LINE1, buf);
+	/* assert EOF */
+	TEST_ASSERT_EQUAL_INT(0, read(fd, buf, sizeof(LINE1) - 1));
+	TEST_ASSERT_EMPTY(buf);
```

Comment:
read() shouldn't modify buff in such case, why empty?

Reply:
This was done to check if the buffer has stayed intact.